[AVR] Replace broken 'avr-unknown-unknown' target with 'avr-unknown-gnu-atmega328' target

dylanmckay opened the pull request:
The `avr-unknown-unknown` target has never worked correctly, always trying to invoke
the host linker and failing. It aimed to be a mirror of AVR-GCC's
default handling of the `avr-unknown-unknown' triple (assume bare
minimum chip features, silently skip linking runtime libraries, etc).
This behaviour is broken-by-default as it will cause a miscompiled executable
when flashed.

This patch improves the AVR builtin target specifications to instead
expose only a 'avr-unknown-gnu-atmega328' target. This target system is
`gnu`, as it uses the AVR-GCC frontend along with avr-binutils. The
target triple ABI is 'atmega328'.

In the future, it should be possible to replace the dependency on
AVR-GCC and binutils by using the in-progress AVR LLD and compiler-rt support.
Perhaps at that point it would make sense to add an
'avr-unknown-unknown-atmega328' target as a better default when
implemented.

There is no current intention to add in-tree AVR target specifications for other
AVR microcontrollers - this one can serve as a reference implementation
for other devices via `rustc --print target-spec-json
avr-unknown-gnu-atmega328p`.

There should be no users of the existing 'avr-unknown-unknown' Rust
target as a custom target specification JSON has always been
recommended, and the avr-unknown-unknown target could never pass the
linking step anyway.

dylanmckay committed:
[AVR] Replace broken 'avr-unknown-unknown' target with 'avr-unknown-gnu-atmega328' target

The `avr-unknown-unknown` target has never worked correctly, always trying to invoke
the host linker and failing. It aimed to be a mirror of AVR-GCC's
default handling of the `avr-unknown-unknown' triple (assume bare
minimum chip features, silently skip linking runtime libraries, etc).
This behaviour is broken-by-default as it will cause a miscompiled executable
when flashed.

This patch improves the AVR builtin target specifications to instead
expose only a 'avr-unknown-gnu-atmega328' target. This target system is
`gnu`, as it uses the AVR-GCC frontend along with avr-binutils. The
target triple ABI is 'atmega328'.

In the future, it should be possible to replace the dependency on
AVR-GCC and binutils by using the in-progress AVR LLD and compiler-rt support.
Perhaps at that point it would make sense to add an
'avr-unknown-unknown-atmega328' target as a better default when
implemented.

There is no current intention to add in-tree AVR target specifications for other
AVR microcontrollers - this one can serve as a reference implementation
for other devices via `rustc --print target-spec-json
avr-unknown-gnu-atmega328p`.

There should be no users of the existing 'avr-unknown-unknown' Rust
target as a custom target specification JSON has always been
recommended, and the avr-unknown-unknown target could never pass the
linking step anyway.

rust-highfive commented:
r? @oli-obk

(rust_highfive has picked a reviewer for you, use r? to override)

petrochenkov commented on src/librustc_target/spec/avr_gnu_base.rs:
```diff
@@ -0,0 +1,34 @@
+            linker: Some("avr-gcc".to_owned()),
+            pre_link_args: vec![(
+                LinkerFlavor::Gcc,
+                vec!["-Os".to_owned(), format!("-mmcu={}", target_cpu)],
```
Does `-Os` have any effect when `gcc` is used as a linker only? What `ld` option does it translate to?

mati865 replied:
At least when it comes to x86 linkers use `-O0` to `-O2` to reduce size, with `-O1` being default.

dylanmckay replied:
I ended up removing it in c4d3e03eb82719f86d20ae105a19c81ce80f5587. Most AVR projects will still need to use a custom JSON target specification file, in which case, it is trivial to add custom link arguments that can be tailored to the project.

petrochenkov commented on src/librustc_target/spec/avr_gnu_base.rs:
```diff
@@ -0,0 +1,34 @@
+            late_link_args: vec![(LinkerFlavor::Gcc, vec!["-lc".to_owned(), "-lgcc".to_owned()])]
+                .into_iter()
+                .collect(),
+            ..super::freestanding_base::opts()
```
Could you inline this and remove `freestanding_base.rs`?
It looks misleadingly generic, but is only used for AVR.

dylanmckay replied:
Agree, fixed in ffe98aa186312f9e8a74765f75e0f32757cc425a

petrochenkov commented on src/librustc_target/spec/avr_gnu_base.rs:
```diff
@@ -0,0 +1,34 @@
+            )]
+            .into_iter()
+            .collect(),
+            late_link_args: vec![(LinkerFlavor::Gcc, vec!["-lc".to_owned(), "-lgcc".to_owned()])]
```
Linking libc is supposed to be done through the `libc` crate ideally.

dylanmckay replied:
Upon further inspection, I don't believe libc is actually required. I copied the options over from the blink example, but it is totally stripped out when linking as there are no references to it in emitted Rust or LLVM code.

I've dropped the dependency in c4d3e03eb82719f86d20ae105a19c81ce80f5587

oli-obk commented:
cc @shepmaster @japaric 

cc @rust-lang/compiler this PR replaces a Tier 3 target that apparently never worked anyway with a controller specific tier 3 target that mostly serves as a reference target for others to copy and reuse

dylanmckay committed:
[AVR] Remove unnecessary arguments passed to the linker for GNU target

In general, linking with libc is not required, only libgcc is needed.
As suggested in the code review, a better option for libc support is by
building it into rust-lang/libc directly.

This also removes the '-Os' argument to the linker, which is a NOP.

dylanmckay committed:
[AVR] Merge the 'freestanding' base target spec into AVR base target spec

The 'freestanding' module was only ever used for AVR. It was an
unnecessary layer of abstraction. This commit merges the
'freestanding_base' module into 'avr_gnu_base'.

dylanmckay committed:
[AVR] Rename the last few remaining references from 'avr-unknown-unknown' to 'avr-unknown-gnu-atmega328'

bors commented:
:umbrella: The latest upstream changes (presumably #75070) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
Sounds good to me, I say land it.

oli-obk commented:
@dylanmckay can you rebase and fix the conflicts? Then we can move to merging this

dylanmckay committed:
[AVR] Replace 'avr-unknown-unknown' with 'avr-unknown-gnu-atmega328' in platform-support.md

dylanmckay commented:
The branch has now been rebased.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit c9ead8c895593452677a229fd19909c83283b33f has been approved by `oli-obk`

<!-- @bors r=oli-obk c9ead8c895593452677a229fd19909c83283b33f -->
<!-- homu: {"type":"Approved","sha":"c9ead8c895593452677a229fd19909c83283b33f","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit c9ead8c895593452677a229fd19909c83283b33f with merge ca9f78dd98835a40aede3b7593cabf425c5bee21...
<!-- homu: {"type":"BuildStarted","head_sha":"c9ead8c895593452677a229fd19909c83283b33f","merge_sha":"ca9f78dd98835a40aede3b7593cabf425c5bee21"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1032436251)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1032436251","builder_name":"checks-actions"} -->

dylanmckay commented:
Strange, seemingly unrelated "command not found" error from bors https://github.com/rust-lang-ci/rust/runs/1032158434#step:23:22604 (`[run-make] run-make-fulldeps\pgo-indirect-call-promotion`)

mati865 commented:
It's spurious failure.

oli-obk commented:
@bors retry

bors commented:
:hourglass: Testing commit c9ead8c895593452677a229fd19909c83283b33f with merge 3d0c847d3353e319ed82598a106e28fd490caa6b...
<!-- homu: {"type":"BuildStarted","head_sha":"c9ead8c895593452677a229fd19909c83283b33f","merge_sha":"3d0c847d3353e319ed82598a106e28fd490caa6b"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1037695414), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36252)
Approved by: oli-obk
Pushing 3d0c847d3353e319ed82598a106e28fd490caa6b to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36252","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1037695414"},"merge_sha":"3d0c847d3353e319ed82598a106e28fd490caa6b"} -->

sunfishcode commented:
Since `-gnu` matches an existing [environment type](https://llvm.org/doxygen/classllvm_1_1Triple.html#a83e907e55fa50e093caa96a0aff96201), does that imply that the `-atmega328` in the triple here refers to a new [object format type](https://llvm.org/doxygen/classllvm_1_1Triple.html#a83e907e55fa50e093caa96a0aff96201)?

dylanmckay commented:
> Since -gnu matches an existing environment type, does that imply that the -atmega328 in the triple here refers to a new object format type?

It does not - here it is only meant to indicate that this triple specificaly targets the GNU toolchain/linker/libraries.
